### PR TITLE
Enrollment code fulfillment module

### PR DESCRIPTION
**Pull request by vkaracic**
PR includes:
- a new fulfillment module for fulfilling enrollment code products
- `OrderLineVouchers` model for recording vouchers created for each order line
- email template for emails sent after an enrollment code order has been fulfilled
- coupon endpoint for generating and downloading a CSV of the order vouchers

Part of https://openedx.atlassian.net/browse/WL-381

Dependencies: https://github.com/edx/ecommerce/pull/712 https://github.com/edx/ecommerce/pull/718

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @vkaracic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=719&repo=edx%2Fecommerce

**Comment by mattdrayer** [comment]
@vkaracic there seems to be an error with the current Travis build for this PR -- I am re-reviewing right now (we already did as part of #702) -- @clintonb when you are ready please jump in as well -- this is the last PR we need to get over the line for the MITPE launch aside from some small theme-related fixes.

File: ecommerce/coupons/views.py [hunk 1]
```diff
@@ -224,3 +232,69 @@ def get(self, request):
...
+                    'title'(str): The title of the product in the line.
+                    'vouchers'(list): List of the vouchers associated to that line.
+                }]
+        """
+        template_name = 'coupons/offer.html'
+
+        try:
+            order = Order.objects.get(number=number)
+        except Order.DoesNotExist:
+            return render(request, template_name, {'error': _('Invalid order.')})
```
Comment (clintonb) [inline ecommerce/coupons/views.py]: `raise Http404`

File: ecommerce/coupons/views.py [hunk 2]
```diff
@@ -224,3 +232,69 @@ def get(self, request):
...
+
+        if request.user != order.user:
+            return render(request, template_name, {'error': _('Invalid order.')})
```
Comment (clintonb) [inline ecommerce/coupons/views.py]: `raise PermissionDenied`

File: ecommerce/coupons/views.py [hunk 3]
```diff
@@ -224,3 +232,69 @@ def get(self, request):
             return HttpResponseRedirect(reverse('basket:summary'))
 
         return HttpResponseRedirect(get_lms_url(''))
+
+
+class EnrollmentCodeCsvView(View):
```
Comment (clintonb) [inline ecommerce/coupons/views.py]: `LoginRequiredMixin`

File: ecommerce/coupons/views.py [hunk 4]
```diff
@@ -224,3 +232,69 @@ def get(self, request):
...
+                }]
+        """
+        template_name = 'coupons/offer.html'
+
+        try:
+            order = Order.objects.get(number=number)
+        except Order.DoesNotExist:
+            return render(request, template_name, {'error': _('Invalid order.')})
+
+        if request.user != order.user:
```
Comment (clintonb) [inline ecommerce/coupons/views.py]: What about staff/super users?

File: ecommerce/coupons/views.py [hunk 5]
```diff
@@ -224,3 +232,69 @@ def get(self, request):
...
+           > Seat in Demo with verified certificate (and ID verification)
+           > Code,Redemption URL
+           > J4HDI5OAUGCSUJJ3,ecommerce.server?code=J4HDI5OAUGCSUJJ3
+           > OZCRR6WXLWGAFWZR,ecommerce.server?code=OZCRR6WXLWGAFWZR
+           > 6KPYL6IO6Y3XL7SI,ecommerce.server?code=6KPYL6IO6Y3XL7SI
+           > NPIJWIKNLRURYVU2,ecommerce.server?code=NPIJWIKNLRURYVU2
+           > 6SZULKPZQYACAODC,ecommerce.server?code=6SZULKPZQYACAODC
+           >
+
+        Args:
```
Comment (clintonb) [inline ecommerce/coupons/views.py]: This is wrong.

File: ecommerce/coupons/views.py [hunk 6]
```diff
@@ -224,3 +232,69 @@ def get(self, request):
...
+        for order_line_voucher in order_line_vouchers:
+            writer.writerow([order_line_voucher.line.product.title])
+            voucher_writer.writeheader()
+
+            for voucher in order_line_voucher.vouchers.all():
+                voucher_writer.writerow({
+                    voucher_field_names[0]: voucher.code,
+                    voucher_field_names[1]: '{url}?code={code}'.format(url=redeem_url, code=voucher.code)
+                })
+            writer.writerow([])
```
Comment (clintonb) [inline ecommerce/coupons/views.py:296]: Why?

File: ecommerce/extensions/fulfillment/modules.py [hunk 7]
```diff
@@ -338,3 +350,112 @@ def revoke_line(self, line):
...
+                'contact_url': get_lms_url('/contact'),
+                'course_name': course_name,
+                'download_csv_link': get_ecommerce_url(reverse('coupons:enrollment_code_csv', args=[order.number])),
+                'enrollment_code_title': lines[0].product.title,
+                'order_number': order.number,
+                'partner_name': order.lines.first().partner.name,
+                'lms_url': get_lms_url(),
+                'receipt_page_url': get_lms_url('{}?orderNum={}'.format(settings.RECEIPT_PAGE_PATH, order.number)),
+            },
+            site=threadlocals.get_current_request().site
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py]: `order.site`

File: ecommerce/extensions/fulfillment/modules.py [hunk 8]
```diff
@@ -338,3 +350,112 @@ def revoke_line(self, line):
...
+
+        send_notification(
+            order.user,
+            'ORDER_WITH_CSV',
+            context={
+                'contact_url': get_lms_url('/contact'),
+                'course_name': course_name,
+                'download_csv_link': get_ecommerce_url(reverse('coupons:enrollment_code_csv', args=[order.number])),
+                'enrollment_code_title': lines[0].product.title,
+                'order_number': order.number,
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py]: Pass the complete `order` object.

File: ecommerce/extensions/fulfillment/modules.py [hunk 9]
```diff
@@ -338,3 +350,112 @@ def revoke_line(self, line):
...
+                'partner_name': order.lines.first().partner.name,
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py]: Pass the complete `partner` object, which should come from `order.site.siteconfiguration.partner`.

File: ecommerce/extensions/fulfillment/modules.py [hunk 10]
```diff
@@ -338,3 +350,112 @@ def revoke_line(self, line):
...
+            line_vouchers = OrderLineVouchers.objects.create(line=line)
+            for voucher in vouchers:
+                line_vouchers.vouchers.add(voucher)
+
+            line.set_status(LINE.COMPLETE)
+
+        # Note (multi-courses): Change from a course_name to a list of course names.
+        course_name = Course.objects.get(id=lines.first().product.attr.course_key).name
+
+        send_notification(
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py]: Move this to a separate method.

File: ecommerce/settings/base.py [hunk 11]
```diff
@@ -516,3 +517,6 @@
 THEME_CACHE_TIMEOUT = 30 * 60
 
 # End Theme settings
+
+# Enrollment codes voucher end datetime
+ENROLLMENT_CODE_EXIPRATION_DATE = datetime.datetime.now() + datetime.timedelta(days=36500)
```
Comment (clintonb) [inline ecommerce/settings/base.py]: `years=100`

File: ecommerce/extensions/fulfillment/modules.py [hunk 12]
```diff
@@ -338,3 +350,112 @@ def revoke_line(self, line):
...
+    def revoke_line(self, line):
+        """ Revokes the specified line.
+
+        Args:
+            line (Line): Order Line to be revoked.
+
+        Returns:
+            True, if the product is revoked; otherwise, False.
+        """
+        raise NotImplementedError("Revoke method not implemented!")
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py:442]: Are there plans to implement this?

File: ecommerce/extensions/fulfillment/tests/test_modules.py [hunk 13]
```diff
@@ -29,6 +35,7 @@
 Applicator = get_class('offer.utils', 'Applicator')
 Benefit = get_model('offer', 'Benefit')
 Catalog = get_model('catalogue', 'Catalog')
+Product = get_model("catalogue", "Product")
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/tests/test_modules.py]: Single quotes

File: ecommerce/extensions/fulfillment/tests/test_modules.py [hunk 14]
```diff
@@ -427,3 +434,45 @@ def test_revoke_line(self):
...
+        course = CourseFactory()
+        course.create_or_update_seat('verified', True, 50, self.partner)
+        enrollment_code = Product.objects.get(product_class__name=ENROLLMENT_CODE_PRODUCT_CLASS_NAME)
+        user = UserFactory()
+        basket = BasketFactory()
+        basket.add_product(enrollment_code, 5)
+        self.order = factories.create_order(number=1, basket=basket, user=user)
+
+    def test_supports_line(self):
+        """Test that a line containing an Enrollment code returns True."""
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/tests/test_modules.py]: What about products with a different product class?

File: ecommerce/extensions/fulfillment/tests/test_modules.py [hunk 15]
```diff
@@ -427,3 +434,45 @@ def test_revoke_line(self):
...
+        line = self.order.lines.first()
+        supports_line = EnrollmentCodeFulfillmentModule().supports_line(line)
+        self.assertTrue(supports_line)
+
+    def test_get_supported_lines(self):
+        """Test that Enrollment code lines where returned."""
+        lines = self.order.lines.all()
+        supported_lines = EnrollmentCodeFulfillmentModule().get_supported_lines(lines)
+        self.assertEqual(len(supported_lines), 1)
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/tests/test_modules.py]: `self.assertEqual(supported_lines, list(lines))`

File: ecommerce/extensions/fulfillment/tests/test_modules.py [hunk 16]
```diff
@@ -427,3 +434,45 @@ def test_revoke_line(self):
...
+        self.assertEqual(supported_lines[0].product.product_class.name, ENROLLMENT_CODE_PRODUCT_CLASS_NAME)
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/tests/test_modules.py]: Drop.

File: ecommerce/extensions/fulfillment/tests/test_modules.py [hunk 17]
```diff
@@ -427,3 +434,45 @@ def test_revoke_line(self):
...
+
+    def test_fulfill_product(self):
+        """Test fulfilling an Enrollment code product."""
+        self.assertEqual(OrderLineVouchers.objects.count(), 0)
+        lines = self.order.lines.all()
+        __, completed_lines = EnrollmentCodeFulfillmentModule().fulfill_product(self.order, lines)
+        self.assertEqual(completed_lines[0].status, LINE.COMPLETE)
+        self.assertEqual(OrderLineVouchers.objects.count(), 1)
+        self.assertEqual(OrderLineVouchers.objects.first().vouchers.count(), 5)
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/tests/test_modules.py]: `5` should be a constant used by this method and `setUp`.

File: ecommerce/extensions/fulfillment/tests/test_modules.py [hunk 18]
```diff
@@ -427,3 +434,45 @@ def test_revoke_line(self):
...
+
+    def test_revoke_line(self):
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/tests/test_modules.py]: Docstring

File: ecommerce/extensions/voucher/models.py [hunk 19]
```diff
@@ -6,5 +6,10 @@ class CouponVouchers(models.Model):
     coupon = models.ForeignKey('catalogue.Product', related_name='coupon_vouchers')
     vouchers = models.ManyToManyField('voucher.Voucher', blank=True, related_name='coupon_vouchers')
 
+
+class OrderLineVouchers(models.Model):
+    line = models.ForeignKey('order.Line', related_name='order_line_vouchers')
+    vouchers = models.ManyToManyField('voucher.Voucher', blank=True, related_name='order_line_vouchers')
```
Comment (clintonb) [inline ecommerce/extensions/voucher/models.py]: `blank=True` implies that you want to create/modify an instance of `OrderLineVouchers`, via Django admin, with no value set for `vouchers`. Why would this ever be true?

File: ecommerce/extensions/voucher/tests/test_utils.py [hunk 20]
```diff
@@ -1,4 +1,5 @@
 import httpretty
+
```
Comment (clintonb) [inline ecommerce/extensions/voucher/tests/test_utils.py]: Remove.

File: ecommerce/extensions/voucher/utils.py [hunk 21]
```diff
@@ -340,16 +341,19 @@ def create_vouchers(
...
-
     vouchers = []
 
-    range_name = (_('Range for {catalog_name}').format(catalog_name=catalog.name))
-    product_range, __ = Range.objects.get_or_create(
-        name=range_name,
-        catalog=catalog,
-    )
+    if catalog is None and _range:
+        # Enrollment code
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Either add more context, or remove this comment. As written, I'm not sure what it means.

**Comment by clintonb** [comment]
Post a screenshot of the email. The logo is wrong.

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.html [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+                        </tr>
+                    </table>
+                </td>
+            <tr>
+
+            <!-- Message Body -->
+            <tr class="cn-content cn-full-width">
+                <td colspan="2">
+                    <p>{% blocktrans %}Dear {{full_name}},{% endblocktrans %}</p>
+                    <p>{% blocktrans %}Thank you for purchasing access to {{course_name}}. Please click <a href="{{download_csv_link}}" target="_blank">HERE</a> to download a CSV file with the enrollment codes for this course. Once you have the codes you can distribute them to your team. {% endblocktrans %}</p>
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.html]: Newlines are perfectly fine.

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.html [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+                    <p>{% blocktrans %}To view your payment information, please visit {{receipt_page_url}}{% endblocktrans %}</p>
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.html]: Add a period at the end of the sentence.

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.html [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+                    <p>{% blocktrans %}To explore other courses, please visit {{lms_url}}{% endblocktrans %}</p>
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.html]: Add a period at the end of the sentence.

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.html [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+                    <p>{% trans "Thank you" %},</p>
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.html]: This stray comma will look weird for some languages. Include the comma for translation, and add a note informing the translators this is for the ending/closing of a message.

``` jinja
{# Translators: This is for the ending/closing of an email to a user. #}
<p>{% trans "Thank you," %}</p>
```

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.txt [hunk 26]
```diff
@@ -0,0 +1,14 @@
+{% load i18n %}
+{% trans "Order confirmation for: " %}{{course_name}}
+
+{% blocktrans %}Dear {{full_name}},{% endblocktrans %}
+
+{% blocktrans %}Thank you for purchasing access to {{course_name}}. Please click <a href="{{download_csv_link}}" target="_blank">HERE</a> to download a CSV file with the enrollment codes for this course. Once you have the codes you can distribute them to your team. {% endblocktrans %}
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.txt]: Text emails cannot have HTML.

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.html [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+                <td style="padding: 20px;">
+                    <table border="0" cellspacing="0" cellpadding="0" width="280" align="left" class="col-2">
+                        <tr>
+                            <td align="left" valign="top" style="font-size:13px; line-height: 20px; color: #aaaaaa;">
+                                {% trans "Order confirmation for" %}:
+                            </td>
+                        </tr>
+                        <tr>
+                            <td align="left" valign="top" style="line-height: 26px; font-size:24px; color: #002D40;">
+                                {{ course_name }}
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.html]: Order confirmations are for orders, not courses.

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.txt [hunk 28]
```diff
@@ -0,0 +1,14 @@
+{% load i18n %}
+{% trans "Order confirmation for: " %}{{course_name}}
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.txt]: Order confirmations are for orders, not courses.

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.txt [hunk 29]
```diff
@@ -0,0 +1,14 @@
...
+
+{% blocktrans %}Dear {{full_name}},{% endblocktrans %}
+
+{% blocktrans %}Thank you for purchasing access to {{course_name}}. Please click <a href="{{download_csv_link}}" target="_blank">HERE</a> to download a CSV file with the enrollment codes for this course. Once you have the codes you can distribute them to your team. {% endblocktrans %}
+
+{% blocktrans %}To explore other courses, please visit {{lms_url}}{% endblocktrans %}
+
+{% trans "Thank you" %},
+
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.txt:9]: Remove this linebreak.

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.txt [hunk 30]
```diff
@@ -0,0 +1,14 @@
...
+{% blocktrans %}The {{partner_name}} team{% endblocktrans %}
+
+{% blocktrans %}{% blocktrans %}You received this message because you purchased enrollment codes for {{course_name}} on {{lms_url}}. If you have any questions, please visit {{contact_url}}.{% endblocktrans %}{% endblocktrans %}
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.txt]: Fix these tags.

**Comment by clintonb** [comment]
Post a copy of the text-based email, too.

File: ecommerce/coupons/views.py (hunk 6 above)
Comment (vkaracic) [inline ecommerce/coupons/views.py:296]: To add a new line between voucher lists, nicer format.

File: ecommerce/settings/base.py (hunk 11 above)
Comment (vkaracic) [inline ecommerce/settings/base.py]: [ `timedelta` ](https://docs.python.org/2/library/datetime.html#datetime.timedelta) does not accept `years` as a parameter.

File: ecommerce/extensions/fulfillment/modules.py (hunk 12 above)
Comment (vkaracic) [inline ecommerce/extensions/fulfillment/modules.py:442]: No that I am aware of.
@mattdrayer @griffresch will order cancellations be supported?

File: ecommerce/settings/base.py [hunk 31]
```diff
@@ -517,6 +518,11 @@
...
 # End Theme settings
 
+
 EDX_DRF_EXTENSIONS = {
     'JWT_PAYLOAD_USER_ATTRIBUTES': ('full_name', 'email', 'tracking_context',),
 }
+
+
+# Enrollment codes voucher end datetime
+ENROLLMENT_CODE_EXIPRATION_DATE = datetime.datetime.now() + datetime.timedelta(days=36500)
```
Comment (clintonb) [inline ecommerce/settings/base.py]: `weeks=520`

More importantly, update your comment to explain why this value is set so far in the future.

File: ecommerce/settings/test.py [hunk 32]
```diff
@@ -125,7 +125,7 @@ def __getitem__(self, item):
 
 # Use production settings for asset compression so that asset compilation can be tested on the CI server.
 COMPRESS_ENABLED = True
-COMPRESS_OFFLINE = True
+COMPRESS_OFFLINE = False
```
Comment (clintonb) [inline ecommerce/settings/test.py]: Revert.

File: ecommerce/templates/customer/emails/commtype_order_with_csv_body.html [hunk 33]
```diff
@@ -0,0 +1,81 @@
...
+                    </table>
+                </td>
+            <tr>
+
+            <!-- Message Body -->
+            <tr class="cn-content cn-full-width">
+                <td colspan="2">
+                    <p>{% blocktrans %}Dear {{full_name}},{% endblocktrans %}</p>
+                    <p>{% blocktrans %}Thank you for purchasing access to {{course_name}}.
+                    Please click <a href="{{download_csv_link}}" target="_blank">HERE</a> to download a CSV file with the enrollment codes for this course.
```
Comment (clintonb) [inline ecommerce/templates/customer/emails/commtype_order_with_csv_body.html]: No need for the `target` attribute in email.

File: ecommerce/settings/base.py [hunk 34]
```diff
@@ -525,4 +525,4 @@
 
 
 # Enrollment codes voucher end datetime
-ENROLLMENT_CODE_EXIPRATION_DATE = datetime.datetime.now + datetime.timedelta(days=36500)
+ENROLLMENT_CODE_EXIPRATION_DATE = datetime.datetime.now() + datetime.timedelta(days=36500)
```
Comment (clintonb) [inline ecommerce/settings/base.py]: This value should be set to the UTC timezone.

File: ecommerce/coupons/views.py [hunk 35]
```diff
@@ -245,23 +247,20 @@ def get(self, request, number):
...
-                    'vouchers'(list): List of the vouchers associated to that line.
-                }]
-        """
-        template_name = 'coupons/offer.html'
+            request (Request): The GET request
+            number (str): Number of the order
 
+        Returns:
+            HttpResponse
+
```
Comment (clintonb) [inline ecommerce/coupons/views.py:258]: Document the exceptions.

File: ecommerce/coupons/views.py [hunk 36]
```diff
@@ -216,3 +225,67 @@ def get(self, request):
             return HttpResponseRedirect(reverse('basket:summary'))
 
         return HttpResponseRedirect(get_lms_url(''))
+
+
+class EnrollmentCodeCsvView(View):
+    """ Download enrollment code CSV file view. """
+
+    @method_decorator(login_required)
```
Comment (clintonb) [inline ecommerce/coupons/views.py:233]: This won't work. Use [`LoginRequiredMixin`](http://django-braces.readthedocs.io/en/latest/access.html#loginrequiredmixin).

File: ecommerce/coupons/views.py [hunk 37]
```diff
@@ -216,3 +225,67 @@ def get(self, request):
...
+           > 6SZULKPZQYACAODC,ecommerce.server?code=6SZULKPZQYACAODC
+           >
+
+        Args:
+            request (Request): The GET request
+            number (str): Number of the order
+
+        Returns:
+            HttpResponse
+
```
Comment (clintonb) [inline ecommerce/coupons/views.py]: Document the exceptions.

File: ecommerce/coupons/views.py (hunk 36 above)
Comment (vkaracic) [inline ecommerce/coupons/views.py:233]: The problem is:

```
>>> import braces
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ImportError: No module named braces
```

File: ecommerce/extensions/fulfillment/modules.py (hunk 12 above)
Comment (mattdrayer) [inline ecommerce/extensions/fulfillment/modules.py:442]: It's my understanding that order cancellations are not a requirement for this iteration, although we may add support for a cancellation workflow in the future.

File: ecommerce/coupons/views.py (hunk 36 above)
Comment (clintonb) [inline ecommerce/coupons/views.py:233]: Odd. Move the decorator to `dispatch` to cover all HTTP verbs.

**Comment by vkaracic** [comment]
Here's a screenshot of the email:
screenshot from 2016-05-13 18-18-39

File: ecommerce/extensions/fulfillment/modules.py (hunk 8 above)
Comment (vkaracic) [inline ecommerce/extensions/fulfillment/modules.py]: Passing a complete object doesn't work here. I've tried with passing `order` and displaying the order number with `{{order.number}}` but it's left blank.

File: ecommerce/extensions/fulfillment/modules.py [hunk 38]
```diff
@@ -459,3 +440,23 @@ def revoke_line(self, line):
...
+        course_name = Course.objects.get(id=product.attr.course_key).name
+        send_notification(
+            order.user,
+            'ORDER_WITH_CSV',
+            context={
+                'contact_url': get_lms_url('/contact'),
+                'course_name': course_name,
+                'download_csv_link': get_ecommerce_url(reverse('coupons:enrollment_code_csv', args=[order.number])),
+                'enrollment_code_title': product.title,
+                'order_number': order,
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py]: This should be order, not order_number.

File: ecommerce/extensions/fulfillment/modules.py [hunk 39]
```diff
@@ -459,3 +440,23 @@ def revoke_line(self, line):
...
+                'partner_name': order.site.siteconfiguration.partner,
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py]: This should be partner, not partner_name.

File: ecommerce/extensions/fulfillment/modules.py [hunk 40]
```diff
@@ -338,3 +349,115 @@ def revoke_line(self, line):
...
+        course_name = Course.objects.get(id=product.attr.course_key).name
+        send_notification(
+            order.user,
+            'ORDER_WITH_CSV',
+            context={
+                'contact_url': get_lms_url('/contact'),
+                'course_name': course_name,
+                'download_csv_link': get_ecommerce_url(reverse('coupons:enrollment_code_csv', args=[order.number])),
+                'enrollment_code_title': product.title,
+                'order_number': order.number,
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py:457]: `'order': order,`

File: ecommerce/extensions/fulfillment/modules.py [hunk 41]
```diff
@@ -338,3 +349,115 @@ def revoke_line(self, line):
...
+                'partner_name': order.site.siteconfiguration.partner.name,
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py:458]: `'partner': order.site.siteconfiguration.partner.name,`

File: ecommerce/extensions/fulfillment/modules.py (hunk 40 above)
Comment (vkaracic) [inline ecommerce/extensions/fulfillment/modules.py:457]: Please see: https://github.com/edx/ecommerce/pull/719#discussion_r63210613

File: ecommerce/extensions/fulfillment/modules.py (hunk 41 above)
Comment (vkaracic) [inline ecommerce/extensions/fulfillment/modules.py:458]: See: https://github.com/edx/ecommerce/pull/719#discussion_r63210613

File: ecommerce/extensions/fulfillment/modules.py [hunk 42]
```diff
@@ -338,3 +349,115 @@ def revoke_line(self, line):
...
+        """ Sends an email with enrollment code order information. """
+        # Note (multi-courses): Change from a course_name to a list of course names.
+        product = order.lines.first().product
+        course_name = Course.objects.get(id=product.attr.course_key).name
+        send_notification(
+            order.user,
+            'ORDER_WITH_CSV',
+            context={
+                'contact_url': get_lms_url('/contact'),
+                'course_name': course_name,
```
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py]: `'course': course`

File: ecommerce/extensions/fulfillment/modules.py (hunk 8 above)
Comment (clintonb) [inline ecommerce/extensions/fulfillment/modules.py]: Your commit was incorrect. You had `'order_number': order`. You should have `'order': order'`.

Pass the full object to make it easier to access other values for future iterations of the message.

Comment (vkaracic) [inline ecommerce/extensions/fulfillment/modules.py]: The tag still doesn't work for the subject. Even though I changed it accordingly.
screenshot from 2016-05-13 19-48-20

**Comment by vkaracic** [comment]
@clintonb I'm reverting back from passing objects to the templates since the object values are not rendered in the TXT templates.
https://github.com/edx/ecommerce/pull/719#discussion_r63223158

**Comment by mattdrayer** [comment]
Yes, we've done several rounds of testing on this trying various approaches -- I will log a ticket to revisit this.

**Comment by mattdrayer** [comment]
@clintonb -- anything else pending on this PR besides the object/context/template thing?

**Comment by clintonb** [comment]
@vkaracic before you revert, let's see the code.

**Comment by vkaracic** [comment]
Here's basically the before and after: https://github.com/edx/ecommerce/pull/719/commits/858c1767b86541129c4b014d8fccb27af36a8a4d

**Comment by clintonb** [comment]
The logo in the email is wrong. I still haven't seen a copy of the text-based email.

**Comment by vkaracic** [comment]
How is the logo wrong?

**Comment by vkaracic** [comment]
Here's the text-based email
screenshot from 2016-05-13 21-01-55

**Comment by vkaracic** [comment]
Sorry about that, that was an image of the the text-based email before the latest revert commit. This is the current one.
screenshot from 2016-05-13 21-05-29

**Comment by clintonb** [comment]
The edX logo is hardcoded. This functionality is not intended for edx.org.

**Comment by mattdrayer** [comment]
@clintonb that is the base email template and the same logo URL that is hardcoded into commtype_credit_receipt_body.html -- we're not actually using this template for MITPE -- there is an edx-themes template override that is being applied which gives us the proper logo.

We can move the logo URL for these two base templates into a configuration setting that could be overridden by another Open edX operator, but it's not impacting the MITPE delivery either way.

**Comment by clintonb** [comment]
Please post a screenshot of the email as it is intended to be delivered (e.g. with proper theming).

**Comment by mattdrayer** [comment]
screen shot 2016-05-13 at 4 26 21 pm

Here's one that I just received after completing a group purchase on 'mtdmicro.sandbox.edx.org'

**Comment by clintonb** [comment]
👍

**Pull request merged**